qemu/tests/cpu_topology_test.py

```diff
@@ -0,0 +1,56 @@
+import random
+
+from virttest import error_context
+from virttest import env_process
+
+from provider.cpu_utils import check_guest_cpu_topology
+
+
+@error_context.context_aware
+def run(test, params, env):
+    """
+    Check guest gets correct vcpu num, cpu cores, processors, sockets, siblings
+
+    1) Boot guest with options: -smp n,cores=x,threads=y,sockets=z...
+    2) Check cpu topology
+
+    :param test: QEMU test object.
+    :param params: Dictionary with test parameters.
+    :param env: Dictionary with the test environment.
+    """
+
+    def check(p_name, exp, check_cmd):
+        """
+        Check the cpu property inside guest
+        :param p_name: Property name
+        :param exp: The expect value
+        :param check_cmd: The param to get check command
+        """
+        res = session.cmd_output(params[check_cmd]).strip()
+        if int(res) != int(exp):
+            test.fail('The vcpu %s number inside guest is %s,'
+                      ' while it is set to %s' % (p_name, res, exp))
+
+    os_type = params['os_type']
+    vcpu_threads_list = [1, 2]
+    if params['machine_type'] == 'pseries':
+        vcpu_threads_list = [1, 2, 4, 8]
+    params['vcpu_cores'] = vcpu_cores = random.randint(1, 10)
+    vm_name = params['main_vm']
+    for vcpu_threads in vcpu_threads_list:
+        vcpu_sockets = min(240 // (vcpu_cores * vcpu_threads),
+                           random.randint(1, 10))
+        vcpu_sockets = 2 if (os_type == 'Windows' and
+                             vcpu_sockets > 2) else vcpu_sockets
+        params['vcpu_sockets'] = vcpu_sockets
+        params['vcpu_threads'] = vcpu_threads
+        params['smp'] = params['vcpu_maxcpus'] = (vcpu_cores *
```

Comment:
In addition, the number of PCPU numbers must be considered. 

The number of vCPUs of a single VM must be less than that of PCPUs. 

Otherwise, softlock up may occur after the VM is started.

Reply:
> In addition, the number of PCPU numbers must be considered.
> 
> The number of vCPUs of a single VM must be less than that of PCPUs.

@nanliu-r Is that true? I suppose not. I am trying to modify it to 'overcommit host CPU number five times' as you mentioned.

Reply:
> > In addition, the number of PCPU numbers must be considered.
> > The number of vCPUs of a single VM must be less than that of PCPUs.
> 
> @nanliu-r Is that true? I suppose not. I am trying to modify it to 'overcommit host CPU number five times' as you mentioned.

If we just check the CPU topology it's no need to overcommit host CPU, does PPC have a different test strategy?

Reply:
If the VM is in an idle state, softlock is not easy to occur.
Since the CPU count in this test case is not the main point, we can restrict the total CPU.
However, on x86, the ACPI ID should less than 255, if you don't want to add `-device intel-iommu,intremap=on,eim=on` and `-M q35,kernel-irqchip=split`

https://github.com/qemu/qemu/blob/93ea484375ab473379dd9c836261ef484bd71ab1/hw/i386/pc.c#L852-L864
```C
    if (x86ms->apic_id_limit > 255 && !xen_enabled()) {
        IntelIOMMUState *iommu = INTEL_IOMMU_DEVICE(x86_iommu_get_default());


        if (!iommu || !x86_iommu_ir_supported(X86_IOMMU_DEVICE(iommu)) ||
            iommu->intr_eim != ON_OFF_AUTO_ON) {
            error_report("current -smp configuration requires "
                         "Extended Interrupt Mode enabled. "
                         "You can add an IOMMU using: "
                         "-device intel-iommu,intremap=on,eim=on");
            exit(EXIT_FAILURE);
        }
    }
}
```

Reply:
Updated to `vcpu_sockets = min(int(host_cpu) * 5 // (vcpu_cores * vcpu_threads), random.randint(1, 10))`, is that okay?

Reply:
@PaulYuuu Please help check this update as well, thanks.

Reply:
Ah, if host_cpu is very small then here `vcpu_sockets` might be 0, let me update it.

Reply:
`vcpu_sockets = min(max(int(host_cpu) * 5 // (vcpu_cores * vcpu_threads), 1), random.randint(1, 10))` for now.